firmware/usbpipe.c

```diff
@@ -34,26 +43,61 @@ void usbpipe_disable() {
     bridge_disable_chan(BRIDGE_USB); // Tells SPI Daemon to close USB Daemon
 }
 
-// Received from USB, send to bridge
-void pipe_usb_out_completion() {
-    if (pipe_state_pc_to_soc == PIPE_WAIT_FOR_USB) {
-        u32 len = usb_ep_out_length(USB_EP_PIPE_OUT);
-        bridge_start_in(BRIDGE_USB, pipe_buffer_pc_to_soc, len);
-        pipe_state_pc_to_soc = PIPE_WAIT_FOR_BRIDGE;
-    } else {
-        invalid();
+void out_ring_step() {
+    // If we are not currently sending
+    // And we haven't filled more packets than we're allowed?
+    // And we have less than a full packet?
```

Comment:
If we are not currently receiving (USB "out" and "in" are always relative to the host)
And there is an empty buffer to receive into
And there isn't an unprocessed short packet -- we can simplify things by only keeping track of the length of one packet, and assuming all others fill the maximum packet size. This means that once we receive a short packet, we have to stop accepting new packets until the short packet is processed. That shouldn't hurt the throughput much because you only send a short packet at the end of a transaction.